src/mctp-bench.c

```diff
@@ -0,0 +1,413 @@
+
+                clock_gettime(CLOCK_MONOTONIC, &recv_ctx.current_time);
+
+                recv_ctx.stats.elapsed_time = get_elapsed_time(&recv_ctx);
+                if (recv_ctx.stats.elapsed_time >= DEFAULT_SECONDS_INTERVAL) {
+                        print_stats(&recv_ctx);
+                        recv_ctx.stats.total_received_len = 0;
+                        recv_ctx.stats.msg_count = 0l;
+                        recv_ctx.stats.msgs_dropped = 0l;
+                        recv_ctx.stats.invalid_payloads = 0l;
+                        clock_gettime(CLOCK_MONOTONIC, &recv_ctx.start_time);
+                }
+        }
+        free(recv_ctx.buf);
+        close(recv_ctx.sd);
+        return 0;
+}
+
+static int mctp_bench_send(struct mctp_bench_send_args send_args)
+{
+        struct sockaddr_mctp addr = {0};
+        struct msg_header *hdr;
+        unsigned char *buf;
+        uint32_t sequence = SEQ_START;
+        int rc, sd, last_rc;
+
+        struct mctp_ioc_tag_ctl2 ctl = {
+            .peer_addr = send_args.eid,
+            .net = send_args.net,
+        };
+
+        sd = socket(AF_MCTP, SOCK_DGRAM, 0);
+        if (sd < 0)
+                err(EXIT_FAILURE, "send: socket");
+
+        addr.smctp_family = AF_MCTP;
+        addr.smctp_network = send_args.net;
+        addr.smctp_addr.s_addr = send_args.eid;
+        addr.smctp_type = 1;
+        printf("send: eid = %d, net = %d, type = %d, msg_len = %zu bytes\n",
+               send_args.eid, send_args.net, addr.smctp_type, send_args.len);
+
+        buf = malloc(send_args.len);
+        if (!buf)
+                err(EXIT_FAILURE, "send: malloc");
+
+        for (size_t i = MSG_HEADER_LEN; i < send_args.len; i++)
+                buf[i] = i & 0xff;
+
+        rc = ioctl(sd, SIOCMCTPALLOCTAG2, &ctl);
+        if (rc)
+                err(EXIT_FAILURE, "send: alloc tag failed");
+
+        hdr = (struct msg_header *)buf;
+        hdr->magic = MAGIC_VAL;
+        while (1) {
+                addr.smctp_tag = ctl.tag;
+                hdr->seq_no = sequence;
+
+                rc = sendto(sd, buf, send_args.len, 0, (struct sockaddr *)&addr,
+                            sizeof(addr));
+                if (rc != (int)send_args.len && rc != last_rc) {
+                        last_rc = rc;
+                        warn("send: sendto(%zd bytes)", send_args.len);
+                }
+
+                sequence++;
+        }
+        free(buf);
+        close(sd);
+        return 0;
+}
+
+static void usage(void)
+{
+        fprintf(stderr, "'mctp-bench send' [len <value>] eid [<net>,]<eid>\n");
+        fprintf(stderr, "'mctp-bench recv'\n");
+}
+
+static int parse_int(char *opt, unsigned int *out)
+{
+        char *endptr;
+
+        errno = 0;
+        *out = strtoul(opt, &endptr, 0);
+        if (endptr == opt || errno == ERANGE) {
+                return -1;
+        }
+        return 0;
+}
+
+static int parse_net_and_eid(struct mctp_bench_send_args *send_args, char *opt)
+{
```

Comment:
We should put these helpers into `mctp-utils.c` eventually (doesn't need to happen in this PR)